### PR TITLE
Redo all the things!?!?!

### PR DESCRIPTION
**Pull request by rwjblue**
* ES6ify (will require major version bump)
* Remove support for legacy template plugins (they already didn't work in ember-cli, so no need to keep support around internally).
* Remove support for `registry.add('<type>', '<some package name>')` in favor of `registry.add('<type>', instanceWithExpectedInterface);`.  I reviewed all usages of `registry.add` in [ember-observer's code search](https://emberobserver.com/code-search?codeQuery=registry.add) and found only [a single instance](https://github.com/BookingSync/ember-cli-i18n-lazy-lookup/blob/master/index.js#L15) of the older string based registration (in an addon without updates for 2 years).
* Flesh out README to make it clearer what are the public API's and add some example usage.
* Add `eslint` (same rules and setup as ember-cli).
* Remove built-in default `minfy-css` plugin. When ember-cli is updated, this will require an addon (similar to `ember-cli-uglify`).

### CODE REVIEW

**Comment by rwjblue** [comment]
> does such an addon exist?

On my local system it does 😉 

> will we have to add it to the Ember CLI blueprints?

Yep.

File: README.md [hunk 1]
```diff
@@ -1,8 +1,62 @@
...
+registry via the `setupPreprocessorRegistry` hook in your addon's `index.js`.
+
+Example:
+
+```js
+module.exports = {
+  name: 'special-js-sauce',
+  
+  setupPreprocessorRegistry(type, registry) {
+    registry.add('js', {
```
Comment (Turbo87) [inline README.md]: might make sense to add `if (type === 'parent')` or something like that to the example

Comment (rwjblue) [inline README.md]: ya, will do

**Comment by Turbo87** [comment]
@rwjblue it seems that this was never released for some reason. do you want to change anything else or should we finally release this as v4.0?

**Comment by rwjblue** [comment]
We can release but I need to finish the clean-css addon before we can update in Ember CLI...

File: index.js [hunk 2]
```diff
@@ -1,109 +1,64 @@
...
-function Registry(plugins, app) {
-  this.registry = {
-    js: [],
-    css: [],
-    'minify-css': [],
-    template: []
-  };
-
-  this.instantiatedPlugins = [];
-  this.availablePlugins = plugins;
```
Comment (Turbo87) [inline index.js]: @rwjblue removing `availablePlugins` unfortunately made this addon incompatible with current Ember CLI releases :-/

File: index.js [hunk 3]
```diff
@@ -1,109 +1,64 @@
...
 
-  var extensions =  registered.reduce(function(memo, plugin) {
-    return memo.concat(plugin.ext);
-  }, [type]).filter(Boolean);
-
-  extensions = require('ember-cli-lodash-subset').uniq(extensions);
-
-  debug('extensions for type %s: %s', type, extensions);
+class Registry {
+  constructor(app) {
```
Comment (Turbo87) [inline index.js]: the `defaultRegistry()` function is still calling this constructor with `new Registry(app.dependencies(), app)`. was that intended?

**Pull request merged**